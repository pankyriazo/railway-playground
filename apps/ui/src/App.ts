import api from "@/api";
import Projects from "./components/Projects";

const App = () => {
  const element = document.createElement("div");
  element.id = "app";
  element.innerHTML = "Loading...";

  api
    .getProjectList("18ee3ebd-813f-4401-a508-f0b7a62c5b9e")
    .then((projectsData) => {
      const projects = new Projects(projectsData.data.me.projects);

      element.innerHTML = "<h1>Projects</h1>";

      element.appendChild(projects.element);
    })
    .catch((error: Error) => {
      element.innerHTML = `Error: ${error.message}`;
    });

  return element;
};

export default App;
