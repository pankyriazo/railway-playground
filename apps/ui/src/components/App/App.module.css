.header {
  background-color: var(--color-bg--inverse);
  color: var(--color-text--inverse);
  font-size: var(--font-size-md);
  font-weight: var(--font-weight-bold);
  padding: var(--spacing-sm) 0;
}
