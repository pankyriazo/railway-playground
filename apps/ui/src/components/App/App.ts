import { Component } from "@/types";
import TokenInput from "@/components/TokenInput";
import Instructions from "@/components/Instructions";
import Projects from "@/components/Projects";

import styles from "./App.module.css";

class App extends Component<DocumentFragment> {
  constructor() {
    super();

    this._element = document.createDocumentFragment();

    this.render();
  }

  override render() {
    if (!this._element) {
      return;
    }

    this._element.textContent = "";

    const header = document.createElement("header");
    header.className = styles.header!;
    const headerInner = document.createElement("div");
    headerInner.className = "container";
    headerInner.innerHTML = `
      <h1 class="title" data-testid="title">
        <a href="/" title="Railway Playground home" data-testid="title-link">
          Railway Playground
        </a>
      </h1>
    `;
    header.appendChild(headerInner);
    this._element.appendChild(header);

    const wrapper = document.createElement("main");
    wrapper.className = "container";

    const input = new TokenInput();
    const inputSection = document.createElement("section");

    const instructions = new Instructions();
    inputSection.appendChild(input.element);
    inputSection.appendChild(instructions.element);

    wrapper.appendChild(inputSection);

    const projects = new Projects();
    const projectsSection = document.createElement("section");
    projectsSection.appendChild(projects.element);
    wrapper.appendChild(projectsSection);

    this._element.appendChild(wrapper);
  }
}

export default App;
