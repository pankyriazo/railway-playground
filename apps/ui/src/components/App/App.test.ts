import { mockComponent, render } from "@/test-helpers";
import App from "./App";

jest.mock("@/components/TokenInput", () => mockComponent("TokenInput"));
jest.mock("@/components/Instructions", () => mockComponent("Instructions"));
jest.mock("@/components/Projects", () => mockComponent("Projects"));

describe("App", () => {
  beforeEach(() => {
    render(App);
  });

  it("should render", () => {
    expect(document.body).toMatchSnapshot();
  });

  describe("header", () => {
    it("should render the correct title", () => {
      const header = document.querySelector("header");

      expect(header).not.toBeNull();
      expect(header!.textContent).toContain("Railway Playground");
    });

    it("should render the correct title link", () => {
      const titleLink = document.querySelector("a");

      expect(titleLink!.getAttribute("href")).toBe("/");
    });
  });

  it("should render the TokenInput component", () => {
    const tokenInput = document.querySelector("[data-component='TokenInput']");

    expect(tokenInput).not.toBeNull();
  });

  it("should render the Instructions component", () => {
    const instructions = document.querySelector(
      "[data-component='Instructions']"
    );

    expect(instructions).not.toBeNull();
  });

  it("should render the Projects component", () => {
    const projects = document.querySelector("[data-component='Projects']");

    expect(projects).not.toBeNull();
  });
});
