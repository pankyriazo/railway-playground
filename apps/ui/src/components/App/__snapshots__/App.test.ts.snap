// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render 1`] = `
<body>
  <header
    class="header"
  >
    <div
      class="container"
    >
      
      
      <h1
        class="title"
        data-testid="title"
      >
        
        
        <a
          data-testid="title-link"
          href="/"
          title="Railway Playground home"
        >
          
          Railway Playground
        
        </a>
        
      
      </h1>
      
    
    </div>
  </header>
  <main
    class="container"
  >
    <section>
      <div
        data-component="TokenInput"
      />
      <div
        data-component="Instructions"
      />
    </section>
    <section>
      <div
        data-component="Projects"
      />
    </section>
  </main>
</body>
`;
